src/main/java/org/gridsuite/filter/server/FilterService.java

```diff
@@ -115,38 +94,40 @@ public FilterService(FiltersToGroovyScript filtersToScript,
         return Optional.empty();
     }
 
-    Optional<AbstractFilterEntity> getFilterEntity(UUID id) {
-        Objects.requireNonNull(id);
-        for (AbstractFilterRepositoryProxy<?, ?> repository : filterRepositories.values()) {
-            Optional<AbstractFilterEntity> res = (Optional<AbstractFilterEntity>) repository.getFilterEntity(id);
-            if (res.isPresent()) {
-                return res;
-            }
-        }
-        return Optional.empty();
-    }
-
     @Transactional
     public <F extends AbstractFilter> AbstractFilter createFilter(F filter) {
-        return filterRepositories.get(filter.getType()).insert(filter);
+        return getRepository(filter).insert(filter);
+    }
+
+    private AbstractFilterRepositoryProxy<? extends AbstractFilterEntity, ? extends FilterRepository<? extends AbstractFilterEntity>> getRepository(AbstractFilter filter) {
+        if (filter.getType().equals(FilterType.SCRIPT)) {
+            return filterRepositories.get(FilterType.SCRIPT.name());
+        }
+        return filterRepositories.get(((FormFilter) filter).getEquipmentFilterForm().getEquipmentType().name());
     }
 
     @Transactional
-    public <F extends AbstractFilter> void changeFilter(UUID id, F filter) {
+    public <F extends AbstractFilter> void changeFilter(UUID id, F newFilter) {
         Optional<AbstractFilter> f = getFilter(id);
         if (f.isPresent()) {
-            if (f.get().getType() == filter.getType()) {  // filter type has not changed
-                filter.setCreationDate(f.get().getCreationDate());
-                filterRepositories.get(filter.getType()).modify(id, filter);
-            } else {  // filter type has changed
-                if ((f.get().getType() == FilterType.SCRIPT && filter.getType() != FilterType.SCRIPT) ||
-                    (f.get().getType() != FilterType.SCRIPT && filter.getType() == FilterType.SCRIPT)) {
-                    throw new PowsyblException(WRONG_FILTER_TYPE);
-                } else {
-                    filterRepositories.get(f.get().getType()).deleteById(id);
-                    filter.setId(id);
-                    filter.setCreationDate(f.get().getCreationDate());
-                    createFilter(filter);
+            if (f.get().getType() == FilterType.SCRIPT && newFilter.getType() == FilterType.SCRIPT) {
+                newFilter.setCreationDate(f.get().getCreationDate());
+                getRepository(newFilter).modify(id, newFilter);
+            } else {
+                FormFilter newFormFilter = (FormFilter) newFilter;
+                FormFilter oldFormFilter = (FormFilter) f.get();
+                if (oldFormFilter.getEquipmentFilterForm().getEquipmentType() == newFormFilter.getEquipmentFilterForm().getEquipmentType()) {  // filter type has not changed
```

Comment:
why differentiate with script ? 
if (f.get().getType() == FilterType.SCRIPT && newFilter.getType() == FilterType.SCRIPT) {
and 
if (oldFormFilter.getEquipmentFilterForm().getEquipmentType() == newFormFilter.getEquipmentFilterForm().getEquipmentType()) )

have the exact same code, can't we test getRepostory(newFilter) == getRepository(f.get()) ? (== as we test reference equality there)
or something like getFilterType(newFilter) instead of getRepository ?

Reply:
Done with equility check on getRepository(). The code is way simpler that way.